feat: add blogpost about JVM upgrade Java 17 to 21 (#9117)
jenkins-infra/jenkins.io: content/blog/2026/05/2026-05-07-JVM-upgrade.adoc

---
layout: post

> title: Java JVM Updates and Memory Improvements at Jenkins

Reviewer: Minor phrasing suggestion.  Optional.
```suggestion
title: Java JVM Updates and Memory Improvements in Jenkins
```

We made use of the link:https://plugins.jenkins.io/monitoring/[Jenkins monitoring plugin] based on Java melody.

The result was clear:

* *Before the upgrade:* monthly mean memory usage of roughly *33–34 GB*
* *After the upgrade:* monthly mean memory usage of roughly *28 GB*

That is a meaningful reduction of approximately *5–6 GB*, or around *15–18%* depending on the baseline.

== Why memory improved

We did not make a single code-level change to achieve this reduction. 
The main driver was the JVM change itself.

Java 21 includes a number of runtime and garbage collection improvements that can positively affect memory behavior, especially in long-running server applications like Jenkins.
While the exact impact depends on workload and configuration, the following areas are commonly relevant:

* improved garbage collection behavior
* runtime optimizations in the JVM

In our case, the upgrade improved the memory usage in our production environment significantly.

== Preparation for bigger updates

. *Measure before and after.* Establish a baseline with monitoring data before the upgrade, then compare metrics after the upgrade to validate improvements.

== Recommendation

> If you are currently running Jenkins on Java 17, it is worth to evaluating, and anyhow needed for the latest LTS Update to upgrade to Java 21.

Reviewer: Minor phrasing suggestion.  Optional.
```suggestion
If you are currently running Jenkins on Java 17, it is worth your time to evaluate the latest LTS update and Java 21.
```
